src/hooks/useWalletCloudBackup.js

```diff
@@ -15,53 +16,93 @@ import logger from 'logger';
 export default function useWalletCloudBackup() {
   const dispatch = useDispatch();
   const { wallets } = useWallets();
-  const { goBack } = useNavigation();
 
   const walletCloudBackup = useCallback(
-    async ({ walletId, password, latestBackup, onError, onSuccess }) => {
-      try {
-        dispatch(setIsWalletLoading(walletLoadingStates.BACKING_UP_WALLET));
+    async ({
+      createBackupFileIfNeeded,
+      handleNoLatestBackup,
+      handlePasswordNotFound,
+      latestBackup,
+      onError,
+      onSuccess,
+      password,
+      walletId,
+    }) => {
+      if (!password && !latestBackup) {
+        // No password, No latest backup meaning
+        // it's a first time backup so we need to show the password sheet
+        handleNoLatestBackup && handleNoLatestBackup();
+        return;
+      }
 
-        let backupFile;
-        if (!latestBackup) {
-          logger.log(
-            'walletCloudBackup:: backing up to icloud',
-            wallets[walletId]
-          );
+      let fetchedPassword = password;
+      if (latestBackup && !password) {
+        // We have a backup but don't have the password, try fetching password
+        fetchedPassword = await fetchBackupPassword();
+      }
 
-          backupFile = await backupWalletToCloud(password, wallets[walletId]);
-        } else {
-          logger.log(
-            'walletCloudBackup:: adding to icloud backup',
-            wallets[walletId],
-            latestBackup
-          );
-          backupFile = await addWalletToCloudBackup(
-            password,
+      // If we still can't get the password, handle password not found
+      if (!fetchedPassword) {
+        handlePasswordNotFound && handlePasswordNotFound();
+        return;
+      }
+
+      dispatch(setIsWalletLoading(walletLoadingStates.BACKING_UP_WALLET));
+
+      // We have the password and we need to add it to an existing backup
+      logger.log('password fetched correctly');
+
+      let updatedBackupFile = null;
+
+      if (createBackupFileIfNeeded && fetchedPassword && !latestBackup) {
+        logger.log('backing up to icloud', wallets[walletId]);
+        updatedBackupFile = await backupWalletToCloud(
+          fetchedPassword,
+          wallets[walletId]
+        );
+      }
+
+      try {
+        if (latestBackup && !updatedBackupFile) {
```

Comment:
I feel like this should be an `else` of the previous `if ( !latestBackup) ` on line 57